File: core/src/main/java/org/predicode/predicator/grammar/NamePrinter.java
package org.predicode.predicator.grammar;

import javax.annotation.Nonnull;


class NamePrinter implements CodePointPrinter {

    @Nonnull
    private final CharSequence name;

    @Nonnull
    private final QuotedName quoted;

    @Nonnull
    private final QuotingStyle quoting;

    @Nonnull
    private final CodePointPrinter print;

    @Nonnull
    private CharClass lastNonSeparating = CharClass.START_CHAR_CLASS;

    NamePrinter(
            @Nonnull CharSequence name,
            @Nonnull QuotedName quoted,
            @Nonnull QuotingStyle quoting,
            @Nonnull CodePointPrinter print) {
        this.name = name;
        this.quoted = quoted;
        this.quoting = quoting;
        this.print = print;
    }

    @Nonnull
    final QuotedName getQuoted() {
        return this.quoted;
    }

    @Nonnull
    final QuotingStyle getQuoting() {
        return this.quoting;
    }

    @Nonnull
    final CharClass getLastNonSeparating() {
        return this.lastNonSeparating;
    }

    @Override
    public void print(int codePoint) {
        this.print.print(codePoint);
    }

    boolean print() {
        //noinspection ResultOfMethodCallIgnored
        this.name.codePoints().boxed().reduce(
                CharClass.START_CHAR_CLASS,
                (cclass, c) -> {

                    final CharClass pclass = cclass.printName(this, c);

                    if (!pclass.separating()) {
                        this.lastNonSeparating = pclass;
                    }

                    return pclass;
                },
                (f, s) -> s);

        if (this.quoting.closeQuote() || !this.lastNonSeparating.nameEnd(this.quoted)) {
            // Close quote if the name does not end with allowed symbol
            print(this.quoted.getQuote());
            return true;
        }

        return false;
    }

}

File: core/src/main/kotlin/org/predicode/predicator/grammar/NamePrinter.kt
package org.predicode.predicator.grammar

internal class NamePrinter(
        val name: CharSequence,
        val quote: CodePoint,
        val quoting: QuotingStyle = QuotingStyle.AUTO_QUOTE,
        private val print: CodePointConsumer) {

    var lastNonSeparating: CharClass = CharClass.START_CHAR_CLASS

    fun print() {
        name.codePoints().mapToObj { it }.reduce(
                CharClass.START_CHAR_CLASS,
                { cclass, c ->
                    cclass.printName(this, c).also {
                        if (!it.separating) lastNonSeparating = it
                    }
                },
                { _, s -> s })
        // Close quote if the name does not end with allowed symbol
        if (quoting.closeQuote || !lastNonSeparating.nameEnd) out(quote)
    }

    fun out(codePoint: Int) = print(codePoint)

}

fun printName(
        name: CharSequence,
        quote: CodePoint,
        quoting: QuotingStyle = QuotingStyle.AUTO_QUOTE,
        print: CodePointConsumer) = NamePrinter(
        name,
        print = print,
        quote = quote,
        quoting = quoting).print()

fun printName(
        name: CharSequence,
        quote: CodePoint,
        quoting: QuotingStyle = QuotingStyle.AUTO_QUOTE): String = StringBuilder().apply {
    printName(name, quote = quote, quoting = quoting) { appendCodePoint(it) }
}.toString()
